### ISSUE TITLE
`avoid_function_literals_in_foreach_calls` does not report with nullable variable

### ISSUE THREAD

**Issue by AlexV525** [dart-lang/linter#3758]
**To Reproduce**

```dart
void main() {
  final list = List.generate(10, (i) => i);
  list.forEach((e) => print(e));
  final List? listNullable = null;
  listNullable?.forEach((e) => print(e)); // <- Not reporting.
}
```

![image](https://user-images.githubusercontent.com/15884415/194694177-ef154628-47a4-40a7-b4be-55df39da70a7.png)

**Comment by AlexV525** [comment]
The rule doesn't apply to map literals too.

**Comment by lrhn** [comment]
The `nullable?.forEach((e) { ... })` kind-of makes sense because you can't just fix it to do:
```dart
for (var e in nullable) { ... }
```
You'd have to introduce a new variable, check if for `null` to get promotion, and then do the `for`/`in`.

At that point, maybe it's not so bad to use `forEach` instead.

It's also correct that it doesn't apply to maps, since `map.forEach` cannot be directly converted to a `for`/`in`, since the callback takes two arguments.

**Comment by AlexV525** [comment]
> The `nullable?.forEach((e) { ... })` kind-of makes sense because you can't just fix it to do:

Maybe we can not raise quick fix for the nullable case then.

> ```
> e.forEach(print);
> ```

I wrote this often, and it seems fine at present.

> It's also correct that it doesn't apply to maps, since `map.forEach` cannot be directly converted to a `for`/`in`, since the callback takes two arguments.

So maybe we can make a better hint by pointing out it only works with "list literals"?

**Comment by srawlins** [comment]
> So maybe we can make a better hint by pointing out it only works with "list literals"?

Not "list literals," but lists, yes. Good idea. I've forked that request to https://github.com/dart-lang/linter/issues/3783.
